src/harness/test_harness_objects.py

```diff
@@ -0,0 +1,324 @@
+    Args:
+      ssl_cert: Path to SSL cert file.
+      ssl_key: Path to SSL key file.
+      testcase: test case object from the caller.
+    """
+    self.ssl_cert = ssl_cert
+    self.ssl_key = ssl_key
+    self.cbsd_objects = {}
+    self.testcase = testcase
+
+  def addCbsdsAndGrants(self, registration_requests, grant_requests):
+    """Construct CBSD object based on the the result of registration and
+    grant requests.
+
+    Args
+      registration_requests: A list of dictionary elements ,where each list elements is
+        individual CBSD registration requests (each of which is itself a dictionary).
+      grant_requests: A list of dictionary elements,where each list elements is
+        individual CBSD grant requests (each of which is itself a dictionary).
+    """
+    # Checking if the number of registration requests matches number of grant requests
+    # There should be exactly one grant request per registration request.
+    assert len(grant_requests) == len(registration_requests)
+    cbsd_ids = self.testcase.assertRegistered(registration_requests)
+
+    # Wrapped version of the grant requests
+    grant_requests_wrap = {
+      'grantRequest': grant_requests
+    }
+
+    #perform grant operation
+    grant_responses = self.testcase._sas.Grant(grant_requests_wrap,
+                                               self.ssl_cert, self.ssl_key)['grantResponse']
+
+    # Make one CBSD object per successful grant request
+    for index, grant_request in enumerate(grant_requests):
+      grant_request["cbsdId"] = cbsd_ids[index]
+      if grant_responses[index]['response']['responseCode'] == 0:
+        cbsd_object = Cbsd(cbsd_ids[index], registration_requests[index],
+                           [grant_responses[index]['grantId']], [grant_request])
+        self.cbsd_objects[cbsd_ids[index]] = cbsd_object
+
+  def getCbsdObjectById(self, cbsd_id):
+      return self.cbsd_objects[cbsd_id]
+
+  def heartbeatRequestForAllActiveGrants(self):
+    """Performs heartbeat request for all grants of CBSD devices at once.
+
+    Returns:
+      heartbeat_requests : list of heart beat requests which are constructed for all the grants
+      by the caller
+      heartbeat_responses: list of heart beat response
+    """
+    # Concatenate all the heartbeat request of CBSD into single request
+    heartbeat_requests = []
+    for cbsd_object_item in self.cbsd_objects.values():
+      heartbeat_requests.extend(cbsd_object_item.constructHeartbeatRequestForAllActiveGrants())
+
+    if len(heartbeat_requests):
+      heartbeat_requests_wrapped = {
+        'heartbeatRequest': heartbeat_requests
+      }
+      print heartbeat_requests_wrapped
+      # Perform heartbeat requests
+      heartbeat_responses = self.testcase._sas.Heartbeat(
+        heartbeat_requests_wrapped, self.ssl_cert, self.ssl_key)['heartbeatResponse']
+
+      for heartbeat_response in heartbeat_responses:
+        cbsd_object = self.cbsd_objects[heartbeat_response['cbsdId']]
```

Comment:
It's safer to use the request instead of the response (in case there was an error with the HBT).

Reply:
Changed   as per suggestion.

Reply:
This change doesn't appear to have been done.